webkit2/jscore.jsc-value.lisp

```diff
@@ -164,3 +164,115 @@
   (value (g-object jsc-value))
   (indent :int))
 (export 'jsc-value-to-json)
+
+(defvar *js-null-value* :null
+  "The value used when translating null from JavaScript to Lisp.")
+(export '*js-null-value*)
+
+(defvar *js-undefined-value* :undefined
+  "The value used when translating undefined from JavaScript to Lisp.")
+(export '*js-undefined-value*)
+
+(defvar *js-false-value* nil
+  "The value used when translating false from JavaScript to Lisp.")
+(export '*js-false-value*)
+
+(defvar *js-true-value* t
+  "The value used when translating true from JavaScript to Lisp.")
+(export '*js-true-value*)
+
+(defvar *js-function-value* :function
+  "The value used when translating function values from JavaScript to Lisp.
+Likely to get deprecated.")
+(export '*js-function-value*)
+
+(defvar *js-array-type* :list
+  "The Lisp data type used when translating arrays from JavaScript to Lisp.")
+(export '*js-array-type*)
+
+(defvar *js-object-type* :hash-table
+  "The Lisp data type used when translating objects from JavaScript to Lisp.")
+(export '*js-object-type*)
+
+(declaim (ftype (function (t &key (:null-value t)
+                             (:undefined-value t)
+                             (:false-value t)
+                             (:true-value t)
+                             (:function-value t)
+                             (:array-type (or null (member :list :vector)))
+                             (:object-type (or null (member :alist :plist :hash-table)))))
+                jsc-value-to-lisp))
+(defun jsc-value-to-lisp (jsc-value &key (null-value *js-null-value*)
+                                      (undefined-value *js-undefined-value*)
+                                      (false-value *js-false-value*)
+                                      (true-value *js-true-value*)
+                                      (function-value *js-function-value*)
+                                      (array-type *js-array-type*)
+                                      (object-type *js-object-type*))
+  "Translate a JSC-VALUE to a lisp value.
+Translates:
+- JS strings to strings.
+- JS numbers to floating numbers.
+- true to TRUE-VALUE (t by default). Also see `*js-true-value*'.
+- false to FALSE-VALUE (nil by default). Also see `*js-false-value*'.
+- null to NULL-VALUE (:null by default). Also see `*js-null-value*'.
+- undefined to UNDEFINED-VALUE (:undefined by default). Also see `*js-undefined-value*'.
+- JS functions to FUNCTION-VALUE (:function by default). This will change in the future!
```

Comment:
Why will it change?

Reply:
Because I want to leverage WebKit to actually call JS from Lisp.

Reply:
So same as the export above, if we are unsure about the API, then I suggest we don't document the return value of the function.
Or maybe document at outside of this list and highlight that's it's an unsupported return value.